Mathlib/Analysis/Analytic/Order.lean

```diff
@@ -192,4 +185,60 @@ theorem isClopen_setOf_order_eq_top (h₁f : AnalyticOnNhd 𝕜 f U) :
     apply (mem_diff w).1
     exact ⟨hw, mem_singleton_iff.not.1 (Subtype.coe_ne_coe.2 h₁w)⟩
 
+/-- On a connected set, there exists a point where a meromorphic function `f` has finite order iff
+`f` has finite order at every point. -/
+theorem exists_order_ne_top_iff_forall (hU : IsConnected U) :
+    (∃ u : U, (hf u u.2).order ≠ ⊤) ↔ (∀ u : U, (hf u u.2).order ≠ ⊤) := by
+  constructor
+  · intro h₂f
+    have := isPreconnected_iff_preconnectedSpace.1 hU.isPreconnected
+    rcases isClopen_iff.1 hf.isClopen_setOf_order_eq_top with h | h
+    · intro u
+      have : u ∉ (∅ : Set U) := by exact fun a => a
+      rw [← h] at this
+      tauto
+    · obtain ⟨u, hU⟩ := h₂f
+      have : u ∈ univ := by trivial
+      rw [← h] at this
+      tauto
+  · intro h₂f
+    obtain ⟨v, hv⟩ := hU.nonempty
+    use ⟨v, hv⟩, h₂f ⟨v, hv⟩
+
+/-- On a preconnected set, a meromorphic function has finite order at one point if it has finite
+order at another point. -/
+theorem order_ne_top_of_isPreconnected {x y : 𝕜} (hU : IsPreconnected U) (h₁x : x ∈ U) (hy : y ∈ U)
+    (h₂x : (hf x h₁x).order ≠ ⊤) :
+    (hf y hy).order ≠ ⊤ :=
+  (hf.exists_order_ne_top_iff_forall ⟨nonempty_of_mem h₁x, hU⟩).1 (by use ⟨x, h₁x⟩) ⟨y, hy⟩
+
+/-- The set where an analytic function has zero or infinite order is discrete within its domain of
+analyticity. -/
+theorem codiscrete_setOf_order_eq_zero_or_top :
+    {u : U | (hf u u.2).order = 0 ∨ (hf u u.2).order = ⊤} ∈ Filter.codiscrete U := by
+  rw [mem_codiscrete_subtype_iff_mem_codiscreteWithin, mem_codiscreteWithin]
+  intro x hx
+  rw [Filter.disjoint_principal_right]
+  rcases (hf x hx).eventually_eq_zero_or_eventually_ne_zero with h₁f | h₁f
+  · filter_upwards [eventually_nhdsWithin_of_eventually_nhds
+      (Filter.Eventually.eventually_nhds h₁f)]
+    simp only [Set.mem_compl_iff, Set.mem_diff, Set.mem_image, Set.mem_setOf_eq, Subtype.exists,
+      exists_and_right, exists_eq_right, not_exists, not_or, not_and, not_forall, Decidable.not_not]
+    intro a _ h₁a
+    use h₁a
+    by_cases h₂a : a = x
+    · rw [← (hf x hx).order_eq_top_iff] at h₁f
+      simp_rw [h₂a]
+      tauto
+    · have : (hf a h₁a).order = ⊤ := by rwa [(hf a h₁a).order_eq_top_iff]
+      tauto
+  · filter_upwards [h₁f]
+    intro a h₁a
+    simp only [Set.mem_compl_iff, Set.mem_diff, Set.mem_image, Set.mem_setOf_eq, Subtype.exists,
+      exists_and_right, exists_eq_right, not_exists, not_or, not_and, not_forall, Decidable.not_not]
+    intro h₂a
+    use h₂a
+    rw [(hf a h₂a).order_eq_zero_iff.2 h₁a]
+    tauto
```

Comment:
```suggestion
  · filter_upwards [eventually_nhdsWithin_of_eventually_nhds h₁f.eventually_nhds] with a ha
    simp +contextual [(hf a _).order_eq_top_iff, ha]
  · filter_upwards [h₁f] with a ha
    simp +contextual [(hf a _).order_eq_zero_iff, ha]
```

Reply:
Same here. I have **never** seen that `+contextual`.

Reply:
It (roughly) means that if you are simplifying `p -> q`, then `simp` will use the fact that `p` is true, while simplifying `q`.